### DISCUSSION TITLE
ReasonML code generation

### DISCUSSION THREAD

**Discussion by yusufyildirim** [prisma/prisma#2478] [Q&A] [answered]
Hey everyone,

I've been watching this project for a long time. Thank you for your effort and your contributions to the open-source community. I actively use ReasonML. For those who don't know what ReasonML is, it's a functional programming language with sound type system developed by Facebook. I'm here today to ask you a question 😄 

## What I need
I think our community (at least some of us) needs an ORM-ish solution and Prisma looks great for this. I want to make Prisma compatible with ReasonML and since Prisma generates JS file and TS declarations, it should be doable without so much hassle. There are some solutions in my mind to do this.

## Potential solutions
1) Write a separate generator (CLI tool) and process Prisma schema or Prisma generated TS file to generate ReasonML bindings.

2) Write a generator for the existing Prisma CLI and add a flag to be able to generate bindings and send PR so everybody can use it without any other tool needed.

3) Do 2. solution but fork Prisma CLI, we don't merge that.

## Sum
I looked at your codebase a bit to see if it would be hard to understand or not. It seemed pretty easy to follow. So, I think 2. solution would be the best but it's your project so I gotta ask you before doing it. What do you guys suggest?

**Comment by janpio** [comment]
Can you share what these "ReasonML bindings" could look like? 
Would that be an additional layer on top of the generated Prisma Client, or a replacement?

**Reply by yusufyildirim** [reply]
It would be some sort of guide to the ReasonML compiler. While compiling ReasonML code to the Javascript, the compiler looks for JS bindings to replace Reason code with them, without any additional runtime cost.

This conversion would be something like this (simplified for the sake of this example):

PrismaClient.re File
```reason
type prismaClient;

[@bs.module "@prisma/client"] [@bs.new]
external make: unit => prismaClient = "PrismaClient";

[@bs.send]
external connect: prismaClient => Promise.t(unit) = "connect";
```

With this binding above, I can use PrismaClient JS file generated by Prisma CLI in my ReasonML project like this:

MyTestFile.re
```reason
let client = PrismaClient.make()
client -> connect();
```

This code would be compiled to this:

```javascript
const PrismaClient = require('@prisma/client').PrismaClient;

const client = new PrismaClient();
client.connect();
```

So as you can see, there is no other layer on top of PrismaClient. Just a bunch of type declarations to make compiler happy 😄 

PS: There might be some cases that we must add some minor runtime cost to make it work or idiomatic with ReasonML though.

**Reply by janpio** [reply]
Interesting.

Can you maybe create a super simple Prisma schema with one model, and then manually write the "ReasonML bindings" that would require to work for you?

You are on the correct path with https://www.prisma.io/docs/reference/tools-and-interfaces/prisma-schema/generators, we also have a spec that explain how to really write a custom generator: https://github.com/prisma/specs/tree/master/generators Depending on what your target output would look like, this could be the way to go here.

**Reply by yusufyildirim** [reply]
Yea I can manually write ReasonML bindings on top of cli generated JS code. That'd work but every time I change the schema and regenerate the Prisma client, I'd have to manually change my bindings also. If I forget to do that, bugs might occur etc. So, it should be automated 😄  Thank you for your help! I'll inform you here if I do that 🙏

**Reply by yusufyildirim** [reply]
`Can you maybe create a super simple Prisma schema with one model, and then manually write the "ReasonML bindings" that would require to work for you?`

I guess I misunderstood you there 😄. Do you want an example project from me to see how it would look like? I'm a bit confused, sorry 😄

**Reply by janpio** [reply]
Haha yes, that is exactly what I was hoping for. We can use that to understand the goal better (without actually knowing ReasonML) and tell you if this is currently possible or not.

**Comment by yusufyildirim** [comment, marked as answer]
I found that there is already a way to easily do this 🤦  https://www.prisma.io/docs/reference/tools-and-interfaces/prisma-schema/generators . It looks like, I can provide generator path in the schema.

**Reply by yusufyildirim** [reply]
Also I find that this guy develops what I need: https://github.com/eWert-Online/prisma-client-reason . It would be great if we make it built-in of course 😄

**Reply by yusufyildirim** [reply]
Hmm, this approach is not exactly what I was looking for but it's okay. I think I can make some workarounds

**Comment by rjhilgefort** [comment]
I tried to pick up where https://github.com/eWert-Online/prisma-client-reason left off, make it compatible with ReScript, as well as bring that project in line with latest generator dependencies. I made some progress, but didn't get all the way there. Had to abandon for now. I put a PR up against that project with my changes if anyone wants to pick it up from there.

https://github.com/eWert-Online/prisma-client-reason/pull/5

**Comment by ludralph** [comment]
Hi @yusufyildirim,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
